src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,139 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
+@SuppressWarnings("unchecked")
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] array;
+    private int size;
+
+    public ArrayList() {
+        array = new Object[DEFAULT_CAPACITY];
+    }
+
+    public ArrayList(int capacity) {
+        if (capacity > 0) {
+            array = new Object[capacity];
+        } else {
+            throw new IllegalArgumentException("Capacity must be greater than zero.");
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        if (size >= array.length) {
+            grow();
+        }
+        array[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Non existent "
+                    + "position: " + index);
+        }
+        if (size >= array.length) {
+            grow();
+        }
+        Object[] result = new Object[++size];
+        copy(array, 0, result, 0, index);
+        result[index] = value;
+        copy(array, index, result, index + 1, size - index - 1);
+        array = result;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (isNotValidIndex(index)) {
+            throw new ArrayListIndexOutOfBoundsException("Not found "
+                    + "element at position: ");
+        } else {
+            return (T) array[index];
+        }
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (isNotValidIndex(index)) {
+            throw new ArrayListIndexOutOfBoundsException("Not found "
+                    + "element at position: " + index);
+        }
+        array[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (isNotValidIndex(index)) {
+            throw new ArrayListIndexOutOfBoundsException("Not found "
+                    + "element at position: " + index);
+        }
+        Object toDelete = array[index];
+        removeAndTrim(index);
+        return (T) toDelete;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = -1;
+        for (int i = 0; i < array.length; i++) {
+            if (array[i] == null ? array[i] == element
+                    : array[i].equals(element)) {
+                index = i;
+                break;
+            }
+        }
+        if (index == -1) {
+            throw new NoSuchElementException("Not found "
+                    + "element at position: " + index);
```

Comment:
The exception message here is misleading. The index is set to -1 when the element is not found, so the message should not reference a position. Consider revising the exception handling to provide a more accurate message.